ThreadSafeWeakPtrControlBlock should retain itself during the object's destruction

kmiller68 opened the pull request:
#### e2a9ba4461a4f1759a8f9afb0a1b7cdda6fc81b8
<pre>
ThreadSafeWeakPtrControlBlock should retain itself during the object&apos;s destruction
<a href="https://bugs.webkit.org/show_bug.cgi?id=284119">https://bugs.webkit.org/show_bug.cgi?id=284119</a>
<a href="https://rdar.apple.com/140988960">rdar://140988960</a>

Reviewed by Mark Lam and Alex Christensen.

We need to take a weak ref to ControlBlocks so they survive until after the
delete of m_object finishes. This comes up when destructors try to eagerly
remove themselves from WeakHashSets. e.g.

```
~MyObject() { m_weakSet.remove(this); }
```

If m_weakSet has the last reference to the ControlBlock then we could end up doing
an amortized clean up, which removes the ControlBlock and destroys it. Then when we
check m_weakSet&apos;s backing table after the cleanup we UAF the ControlBlock.

* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeWeakPtrControlBlock::strongDeref const):

Canonical link: <a href="https://commits.webkit.org/287425@main">https://commits.webkit.org/287425@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5457d50aef79152cc9ce7f2bcf14d0934b06780

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79612 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58608 "") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33015 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84164 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30669 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67709 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6902 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62243 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20096 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52303 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72526 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42551 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49640 "Failed to checkout and rebase branch from PR 37503") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26677 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29093 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72670 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70784 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85576 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78750 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6867 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4792 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70488 "Passed tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7038 "") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68371 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69733 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17392 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12666 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101089 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6818 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24668 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6715 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10201 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8515 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -70,23 +70,33 @@ class ThreadSafeWeakPtrControlBlock {
+            // We retained ourselves above.
+            if (!(--m_weakReferenceCount)) {
+                // release the lock now so we don't do it in Locker's destuctor after we've already called delete.
+                locker.unlockEarly();
```
I think it would be a bit nicer for locker to just go out of scope, but this works too I guess.

kmiller68 replied:
Done.

MenloDorian reviewed (approved):
r=me too

kmiller68 committed:
ThreadSafeWeakPtrControlBlock should retain itself during the object's destruction
https://bugs.webkit.org/show_bug.cgi?id=284119
rdar://140988960

Reviewed by Mark Lam and Alex Christensen.

We need to take a weak ref to ControlBlocks so they survive until after the
delete of m_object finishes. This comes up when destructors try to eagerly
remove themselves from WeakHashSets. e.g.

```
~MyObject() { m_weakSet.remove(this); }
```

If m_weakSet has the last reference to the ControlBlock then we could end up doing
an amortized clean up, which removes the ControlBlock and destroys it. Then when we
check m_weakSet's backing table after the cleanup we UAF the ControlBlock.

* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeWeakPtrControlBlock::strongDeref const):

Canonical link: https://commits.webkit.org/287425@main
